hibernate-core/src/main/java/org/hibernate/query/criteria/internal/path/PluralAttributePath.java

```diff
@@ -44,39 +46,75 @@ private CollectionPersister resolvePersister(CriteriaBuilderImpl criteriaBuilder
-			// the attribute is declared in a mappedsuperclass
-			if ( getPathSource().getModel().getBindableType() == Bindable.BindableType.ENTITY_TYPE ) {
-				// the role will be assigned to the "nearest" EnityType subclass of the MappedSuperclassType
-				EntityType entityTypeNearestDeclaringType = (EntityType) getPathSource().getModel();
-				IdentifiableType superType = entityTypeNearestDeclaringType.getSupertype();
-				IdentifiableType previousType = entityTypeNearestDeclaringType;
-				while ( superType != attribute.getDeclaringType() ) {
-					if ( superType == null ) {
-						throw new IllegalStateException(
-								String.format(
-									"Cannot determine nearest EntityType extending mapped superclass [%s]; [%s] extends [%s], but supertype of [%s] is null",
-										attribute.getDeclaringType().getJavaType().getName(),
-										( (EntityType) getPathSource().getModel() ).getJavaType().getName(),
-										previousType.getJavaType().getName(),
-										previousType.getJavaType().getName()
-								)
-						);
-					}
-					if ( superType.getPersistenceType() == Type.PersistenceType.ENTITY ) {
-						entityTypeNearestDeclaringType = (EntityType) superType;
+		switch ( attribute.getDeclaringType().getPersistenceType() ) {
+			case ENTITY: {
+				return attribute.getDeclaringType().getJavaType().getName() + '.' + attribute.getName();
+			}
+			case MAPPED_SUPERCLASS: {
+				// the attribute is declared in a mappedsuperclass
+				if ( getPathSource().getModel().getBindableType() == Bindable.BindableType.ENTITY_TYPE ) {
+					// the role will be assigned to the "nearest" EnityType subclass of the MappedSuperclassType
+					EntityType entityTypeNearestDeclaringType = (EntityType) getPathSource().getModel();
+					IdentifiableType superType = entityTypeNearestDeclaringType.getSupertype();
+					IdentifiableType previousType = entityTypeNearestDeclaringType;
+					while ( superType != attribute.getDeclaringType() ) {
+						if ( superType == null ) {
+							throw new IllegalStateException(
+									String.format(
+											"Cannot determine nearest EntityType extending mapped superclass [%s]; [%s] extends [%s], but supertype of [%s] is null",
+											attribute.getDeclaringType().getJavaType().getName(),
+											( (EntityType) getPathSource().getModel() ).getJavaType().getName(),
+											previousType.getJavaType().getName(),
+											previousType.getJavaType().getName()
+									)
+							);
+						}
+						if ( superType.getPersistenceType() == Type.PersistenceType.ENTITY ) {
+							entityTypeNearestDeclaringType = (EntityType) superType;
+						}
+						previousType = superType;
+						superType = superType.getSupertype();
 					}
-					previousType = superType;
-					superType = superType.getSupertype();
+					return entityTypeNearestDeclaringType.getJavaType().getName() + '.' + attribute.getName();
+				}
+				else {
```

Comment:
I am not absolutely sure that it is appropriate to throw an exception here. There are no test failures due to this.